### ISSUE TITLE
Why do 3d VAE suddenly appear Mosaic in training

### ISSUE THREAD

**Issue by guanzhenghua** [mueller-franzes/medfusion#10]
Thank you so much for your paper and code.
I tried VAE model in training in 2d images and the results were very good. However, I encountered some difficulties in training 3D mri. I don't know why. When training VAE in 3D mri, mosaics are generated suddenly.
Here is the thing. I input VAE model with mri of 128*128*128 size, and take pictures for every 200 samples trained. 
The first sample was very blurry but barely recognizable as a brain image. the loss is about 2.0e+7, The picture is as follows.
![1](https://user-images.githubusercontent.com/114739026/231953387-0867fd88-0305-4df2-bfb9-d80154cb5669.jpg)
At the 27th sample, can get a clearer picture of the brain, the loss is about 2.0e+6, The picture is as follows.
![27](https://user-images.githubusercontent.com/114739026/231953418-2401e1c2-4c74-4b3c-931f-4c67dfebb177.jpg)
But on the next sample, the 28th sample, get a bunch of mosaics, and all the subsequent samples are mosaics, The losses are higher than they were in the first place. The picture is as follows.
![28](https://user-images.githubusercontent.com/114739026/231953435-c0f51211-d269-4028-b8db-09f13b36f528.jpg)
This problem has been bothering me for a long time. Thank you very much for your help！

**Comment by mueller-franzes** [comment]
Unfortunately, I do not know for sure what is causing this error. It seems that there is an instability in the optimization of the loss function. Are you training with 32bit or 16?  What learn rate are you using (`optimizer_kwargs={'lr':1e-4}`)? Another cause could be the correct weighting between embedding loss (KL) and reconstruction (adjustable via `embedding_loss_weight`). 

But first I would try to uncomment the SSIM loss (see [self.ssim_loss(pred, target)](https://github.com/mueller-franzes/medfusion/blob/4b92854dd1599edc091f9793390be4dea07b62e3/medical_diffusion/models/embedders/latent_embedders.py#L807) ). It could be that it is responsible for the instability.

**Comment by guanzhenghua** [comment]
Thank you very much for your thoughtful reply.
I've solved the problem with the Mosaic. Maybe it was the 3d mri, so I used the smaller lr : 5e-6 and it didn't happen.
But there's a new problem. 3d VAE doesn't work well. 
This is the result of my 50 iterations of 3000 mri each. First is the original. The edges are very clear. The second is a reconstruction, but with blurred edges.

So the question is I don't know how to adjust the parameters, for exmaple lr ，embedding_loss_weight ，number of VAE layers(maybe dimensions are reduced by a factor of eight is too much for 3d ,maybe four can work better).or VAE do not have the ability to reconstruct 3d mri , I need to choose another model. So I sincerely hope to get your advice. Thank you very much for your reply
![1](https://user-images.githubusercontent.com/114739026/232657867-3b8a7821-2cf5-4dea-9dcf-8101babfabf2.jpg)
![2](https://user-images.githubusercontent.com/114739026/232657912-b4577cca-6c97-46cd-a282-8cbf66cc960f.jpg)

**Comment by mueller-franzes** [comment]
Glad to hear you were able to solve the first problem :) 

The easiest way to improve the quality is to reduce the compression. Two things are crucial here
1) `emb_channels=8` (increase for better quality)
2) `strides = [ 1, 2, 2, 2]` (reduce for better quality) - if your image is not isotropic e.g. 32x512x512 I would try the following e.g. [ 1, (1, 2, 2), (1, 2, 2), 2].

Might also help: 
`embedding_loss_weight=0`

Finally, if you have enough memory, it helps to increase the channels (the effect is rather small):
`hid_chs = [ 64, 128, 256, 512]`

**Comment by guanzhenghua** [comment]
I'm sorry that I didn't see your reply until two days later.
Thank you very much for answering each question in such detail.
First of all, I'm very sorry. I forgot to state that the size of the image is 128*128*128.
In the second question, which was two days ago, I asked how to reconstruct high-quality images with 3d mri. Here are the parameters I selected at that time. Very coincidentally, many of the parameters are the same as you suggested.

1.emb_channels=8   
2.strides = [ 1, 2, 2, 1]  
3.embedding_loss_weight= 1e-5
4.hid_chs = [ 64, 128, 256, 512]  (my gpu is nvidia RTX A6000 48g )

 I would like to ask three more questions here.
1, What is the purpose of embedding_loss_weight, is it necessary to reduce embedding loss when reducing loss?
2. Whether need to train for long enough, in this paper Medical Diffusion Denoising Diffusion Probabilistic Models for 3D Medical Image Generation, trained for a week, Maybe the quality I'm worried about is not good because I only trained for three days.
3. Finally, and most important, There are four models in your code，VQVAE, VQGAN, VAE, VAEGAN, which one is better for 3d image,in this paper Medical Diffusion Denoising Diffusion Probabilistic Models for 3D Medical Image Generation, choose vqgan.
I found out you were one of the authors, there express my sincerely admiration. i wonder if vagan is the beat model for 3D medical image, and i not reconstruct high-quality images beacuse of  used VAE model.
Thank you very much for your reply.

**Comment by mueller-franzes** [comment]
1.)  
For VAE `embedding_loss_weight` corresponds to the weighting between KL loss and reconstruction (eg. L2). 

I don't think it is beneficial to reduce embedding loss when the loss decreases but there is some discussion about the optimal determination.  
[Discussion](https://stats.stackexchange.com/questions/332179/how-to-weight-kld-loss-vs-reconstruction-loss-in-variational-auto-encoder)

I have orientated myself on stable diffusion: 
[Stable-Diffusion implementation](https://github.com/CompVis/stable-diffusion/blob/21f890f9da3cfbeaba8e2ac3c425ee9e998d5229/ldm/modules/losses/contperceptual.py#L83)
[Stable-Diffusion config](https://github.com/CompVis/stable-diffusion/blob/21f890f9da3cfbeaba8e2ac3c425ee9e998d5229/configs/autoencoder/autoencoder_kl_64x64x3.yaml#L11)

It even seemed advantageous to me to set `embedding_loss_weight=0` for the VAE. 

For VQVAE `embedding_loss_weight` corresponds to the weighting between the codebook/Embedding loss [link](https://github.com/mueller-franzes/medfusion/blob/4b92854dd1599edc091f9793390be4dea07b62e3/medical_diffusion/models/embedders/latent_embedders.py#L67) and reconstruction (eg. L2). 

2.) Yes, the training can take a long time, have you looked at the tensorboard metrics (e.g. SSIM)? Are they still decreasing or increasing after three days? 

3.) In my experience VAE seemed to work just as well as VQVAE. If you try both, let me know which worked better for you, I'd be interested. 

Oh, that's an important point I forgot to mention- sorry! I could see a jump in the quality of the histology images after using VAE-GAN. The GAN is simply another term in the loss function. You can load your pre-trained VAE and continue training with VAEGAN. It's similar to stable diffusion, the GAN loss is "switched on" after x-steps [see here](https://github.com/CompVis/stable-diffusion/blob/21f890f9da3cfbeaba8e2ac3c425ee9e998d5229/configs/autoencoder/autoencoder_kl_32x32x4.yaml#L10)
